### PR TITLE
libmysqlclient: use version range for openssl + improve robustness of openssl discovery & link

### PR DESCRIPTION
**Pull request by SpaceIm**
- Use version range for openssl
- As a consequence, a patch is applied to upstream CMakeLists to be robust to transitive dependencies of OpenSSL (zlib was disabled in openssl 1.1.x recipe so this issue was hidden).

---

- [ ] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [ ] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
```
/usr/bin/ld: /home/conan/w/prod/BuildSingleReference/.conan/data/openssl/3.1.0/_/_/package/3b92fb92d86b5880d7fcd3124cd3e37e04f7e25f/lib/libcrypto.a(libcrypto-lib-c_zlib.o): in function `zlib_stateful_expand_block':
c_zlib.c:(.text+0x51): undefined reference to `inflate'
```

Likely a bad way to link OpenSSL in CMakeLists of libmysqlclient.

**Comment by RubenRBS** [comment]
Quick question: What's the reasoning behind only adding 3.0 for the newer versions?

Also quick heads up: We're trying to come up with a way to better manage openssl specifically for all its dependents, so in the future we don't need to open 50PRs bumping their versions, so stay tuned for more :)

**Comment by SpaceIm** [comment]
> Quick question: What's the reasoning behind only adding 3.0 for the newer versions?

Because libmysqlclient is not compatible with openssl 3.x before 8.0.30

File: recipes/libmysqlclient/all/conanfile.py [hunk 1]
```diff
@@ -109,8 +112,7 @@ def _cmake_new_enough(self, required_version):
 
     def build_requirements(self):
         if is_apple_os(self) and not self._cmake_new_enough("3.18"):
-            # CMake 3.18 or higher is required if Apple, but CI of CCI may run CMake 3.15
-            self.tool_requires("cmake/3.25.3")
+            self.tool_requires("cmake/[^3.18]")
```
Comment (RubenRBS) [inline recipes/libmysqlclient/all/conanfile.py:115]: Can you fix up this range too please? :)

**Review by RubenRBS — CHANGES_REQUESTED** [review]

File: recipes/libmysqlclient/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/libmysqlclient/all/conanfile.py:115]: Sorry I don't understand. It's fixed already.

**Comment by SpaceIm** [comment]
By the way, these `replace_in_file()` with strict=False should not be allowed in conancenter, they hide whether they are still required or not for versions currently maintained.

**Review by jwillikers — APPROVED** [review]

**Review by MartinDelille — APPROVED** [review]

File: recipes/libmysqlclient/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/libmysqlclient/all/conanfile.py]: @RubenRBS I guess you meant that there should be an upper bound? It's fixed now.

**Comment by jcar87** [comment]
Thanks for your contribution @SpaceIm  - could you please update the PR description to motivate the changes related to "Fix discovery & link to OpenSSL" ? Thanks

**Comment by SpaceIm** [comment]
The fix is a consequence of openssl bump (because now zlib is a transitive dependency), not the opposite, but I can update the title if you want. It will trigger a new build.

**Comment by jcar87** [comment]
> The fix is a consequence of openssl bump (because now zlib is a transitive dependency), not the opposite, but I can update the title if you want. It will trigger a new build.

I mean the description sorry  - I'm expecting an explanation of what the problem was and how this is solved by this. We have a high volume of pull requests that we are trying to prioritise for this week and the next - it helps massively if all the information required to review is in the description.

File: recipes/libmysqlclient/all/conanfile.py (hunk 1 above)
Comment (RubenRBS) [reply recipes/libmysqlclient/all/conanfile.py]: Yeah sorry, I should have made myself more clear!

**Review by RubenRBS — APPROVED** [review]

**Pull request merged**